README
= MouseTracker =

== Features ==

* Tracks mouse using sound (pan for x coordenate, frequency for y coordenate)
* Daemon can be controled via d-bus interface and uses d-bus activation
* Writen in python
* Uses gstreamer for sound output
* uses gtk, gdk and pyatspi for capturing mouse events
* Stores preferences in json format


== requirements ==

* pygtk 2.0
* python gstreamer bindings (pygst) 0.10
* pyatspi
* dbus-python
* python >= 2.4 and simplejson module or python >= 2.6




ChangeLog
use ``git log`` to see the logs



src/mousetracker/settings.py
import json
import os.path

import logging

_log = logging.getLogger("mousetracker.settings")

class Settings(object):

    def __init__(self, config_file=None):
        if config_file is None:
            self._config_file = self._getDefaultConfigFile()
        else:
            self._config_file = config_file

        _log.debug("Reading configuration from %s" % self._config_file)
        try:
            with open(self._config_file, "r") as f:
                self._config = json.load(f)
        except IOError, e:
            _log.exception(e, "error reading config file")
            self._config = {}

    def flush(self):
        _log.debug("flushing configuration to disk")
        with open(self._config_file, "w") as f:
            json.dump(self._config, f)

    def _getDefaultConfigFile(self):
        home = os.path.expanduser("~")
        xdbg_config_path = ".config/mousetracker"
        return os.path.join(home, xdbg_config_path)

    def __getitem__(self, key):
        return self._config[key]

    def __setitem__(self, key, value):
        self._config[key] = value

    def __delitem__(self, key):
        self._config[key]

    def __contains__(self, key):
        return key in self._config


src/mouselocator/mouselocator.py
#! /usr/bin/env python

import pygtk
pygtk.require("2.0")

import gobject
import gtk
import pyatspi

import pygst
pygst.require("0.10")

import gst



class MouseSoundLocator(object):
    """ Mouse locator class to locate the mouse using sound """

    def __init__(self):
        self._running = False
        
        # construct pipeline
        self._player = gst.Pipeline()
        self._source = gst.element_factory_make("audiotestsrc")
        self._source.set_property("num_buffers", 5)
        self._source.set_property("wave", 3)
        self._panorama = gst.element_factory_make("audiopanorama")
        self._converter = gst.element_factory_make("audioconvert")
        self._sink = gst.element_factory_make("gconfaudiosink")
        self._player.add(self._source, self._panorama, self._converter, self._sink)
        gst.element_link_many(self._source, self._panorama, self._converter, self._sink)
        bus = self._player.get_bus()
        bus.add_signal_watch()
        bus.connect("message", self._onMessage)

    def run(self):
        if self._running:
            return
        pyatspi.Registry.registerEventListener(self._onMouseMoved, "mouse:abs")
        self._running = True

    def stop(self):
        if not self._running:
            return
        pyatspi.Registry.deregisterEventListener(self._onMouseMoved, "mouse:abs")
        self._running = False

    def _onMouseMoved(self, event):
        x, y = event.detail1, event.detail2
        freq, pan = self._calculateSoundParametersFromMousePosition(x, y)
        self._beep(freq, pan)

    def _calculateSoundParametersFromMousePosition(self, x, y):
        screen = gtk.gdk.screen_get_default()
        pan = 2 * (float(x) / screen.get_width())  -1
        freq = 200 + (screen.get_height() - y) * 1000 / screen.get_height()
        return (freq, pan)

    def _beep(self, frequency, panorama):
        self._player.set_state(gst.STATE_READY)
        self._source.set_property("freq", frequency)
        self._panorama.set_property("panorama", panorama)
        self._player.set_state(gst.STATE_PLAYING)

    def _onMessage(self, bus, message):
        t = message.type
        if t == gst.MESSAGE_EOS:
            self._player.set_state(gst.STATE_NULL)
        elif t == gst.MESSAGE_ERROR:
            self._player.set_state(gst.STATE_NULL)
            err, debug = message.parse_error()
            print "%s: %s" %(err, debug)


class Application(object):

    def __init__(self):
        pass


    def main(self, argv):

    # setup signal handlers
        import signal
        signal.signal(signal.SIGINT, lambda signum, stackframe : self.quit())

        # startup sound locator
        self._locator = MouseSoundLocator()
        self._locator.run()
        pyatspi.Registry.start()

    def quit(self):
        self._locator.stop()
        pyatspi.Registry.stop()

if __name__ == '__main__':
    import sys
    Application().main(sys.argv)


src/mousetracker/mousetracker.py
#! /usr/bin/env python

import sys

import gobject
import pyatspi

import dbus
import dbus.service
import dbus.mainloop
import dbus.glib

import logging

from trackers import MousePositionTracker

import settings

BUS_NAME = "mousetracker.Daemon"
dbus.mainloop.glib.DBusGMainLoop(set_as_default=True)

_log = logging.getLogger("mousetracker")
_handler = logging.StreamHandler()
_log.addHandler(_handler)
_log.setLevel(logging.DEBUG)

class Daemon(object):

    def __init__(self, args=[]):
    
    # read configuration
                self._settings = settings.Settings()
                
                # store reference to session bus
                self._bus = dbus.SessionBus()
                self._bus_name = None

    def main(self):
        # setup signal handlers
        import signal
        signal.signal(signal.SIGINT, lambda signum, stackframe : self.quit())

        # verfify if there is another instance of the daemon running
        if self._busNameInUse():
            _log.critical("Bus name already in use, another daemon instance is running on your system. Exiting now.")
            sys.exit(-1)
        # store reference to bus_name
        self._bus_name = dbus.service.BusName(BUS_NAME, bus=dbus.SessionBus())

        # startup sound tracker
        kwargs = {}
        if 'positiontracker' in self._settings:
            kwargs = self._settings['positiontracker']
        self._tracker = MousePositionTracker(**kwargs)
        self._tracker.running = True
        
        _log.info("Starting Daemon")
        pyatspi.Registry.start()

    def quit(self):
        self._tracker.running = False
        self._settings['positionTracker'] = self._tracker.getAll()
        self._settings.flush()
        pyatspi.Registry.stop()
        _log.info("Daemon stoped")

    def _busNameInUse(self):
        daemon_proxy = self._bus.get_object(dbus.BUS_DAEMON_NAME, dbus.BUS_DAEMON_PATH)
        return daemon_proxy.NameHasOwner(BUS_NAME, dbus_interface=dbus.BUS_DAEMON_IFACE)

if __name__ == '__main__':
    Daemon(sys.argv).main()


src/mousetracker/trackers.py
"""
Mouse trackers

For now it has just the BeepsMouseTracker
"""

import pygtk
pygtk.require("2.0")
import gtk

import pygst
pygst.require("0.10")
import gst

import gobject
import pyatspi

import dbus
import dbus.service
import dbus.gobject_service

import logging

class TrackerPropertyException(dbus.DBusException):
    _dbus_error_name="mousetracker.tracker.TrackerPropertyException"

class MouseTracker(dbus.gobject_service.ExportedGObject):
    _log = logging.getLogger("mousetracker.tracker")
    _log.setLevel(logging.DEBUG)

    PROPERTIES_IFACE = "mousetracker.tracker.Properties"

    def __init__(self, **kwargs):
        from daemon import BUS_NAME
        dbus.gobject_service.ExportedGObject.__init__(self, conn=dbus.SessionBus(),
            bus_name=dbus.service.BusName(BUS_NAME, bus=dbus.SessionBus()), **kwargs)
        self.connect("notify", self._onPropertyChanged)

    @dbus.service.method(PROPERTIES_IFACE,
    in_signature="",
    out_signature="a{sv}")
    def getAll(self):
        dict = {}
        for prop in self.props:
            dict[prop.name] = self.get_property(prop.name)
        self._log.debug("getAll method called. Properties are %s" % str(dict))
        return dict

    @dbus.service.method(PROPERTIES_IFACE,
in_signature="s",
out_signature="v")
    def get(self, name):
        try:
            value = self.get_property(name)
            self._log.debug("Get %s=%s." %(name, value))
            return value
        except TypeError, e:
            self._log.exception(e)
            raise TrackerPropertyError(e.message)

    @dbus.service.method(PROPERTIES_IFACE,
    in_signature="sv",
    out_signature="")
    def set(self, name, value):
        try:
            self._log.debug("Set %s=%s." %(name, value))
            self.set_property(name, value)
        except TypeError, e:
            self._log.exception(e)
            raise TrackerPropertyError(e.message)

    def _onPropertyChanged(self, object, spec, userdata=None):
        name = spec.name
        value = self.get_property(name)
        self.propertyChanged(name, value)

    @dbus.service.signal(PROPERTIES_IFACE,
    signature="sv")
    def propertyChanged(self, name, value):
        self._log.debug("Notifying of %s changing to %s" % (name, value))
        return (name, value)

class MousePositionTracker(MouseTracker):
    """ Mouse locator class to locate the mouse using sound """
    object_path = "/mousetracker/tracker/MousePositionTracker"

    def __init__(self, **kwargs):
        self._running = False
        MouseTracker.__init__(self, object_path=MousePositionTracker.object_path, **kwargs)
        # construct pipeline
        self._player = gst.Pipeline()
        self._source = gst.element_factory_make("audiotestsrc")
        self._source.set_property("num_buffers", 5)
        self._source.set_property("wave", 3)
        self._panorama = gst.element_factory_make("audiopanorama")
        self._converter = gst.element_factory_make("audioconvert")
        self._sink = gst.element_factory_make("gconfaudiosink")
        self._player.add(self._source, self._panorama, self._converter, self._sink)
        gst.element_link_many(self._source, self._panorama, self._converter, self._sink)
        bus = self._player.get_bus()
        bus.add_signal_watch()
        bus.connect("message", self._onMessage)

    def _get_volume(self):
        return self._source.get_property("volume")

    def _set_volume(self, volume):
        self._source.set_property("volume", volume)

    volume = gobject.property(
        type=float,
        nick='volume',
        default=0.5,
        minimum=0.0,
        maximum=1.0,
        getter=_get_volume,
        setter=_set_volume
        )

    max_freq = gobject.property(
        type=int,
        nick='max_freq',
        default=1500,
        minimum=600,
        maximum=2000
        )

    min_freq = gobject.property(
        type=int,
        nick='min_freq',
        default=200,
        minimum=20,
        maximum=500
        )

    def _get_running(self):
        try:
            return self._running
        except AttributeError:
            return False

    def _set_running(self, value):
        if value != self._running: # if it is different
            if value:
                self._run()
            else:
                self._stop()
            self._running = value

    def _run(self):
        pyatspi.Registry.registerEventListener(self._onMouseMoved, "mouse:abs")


    def _stop(self):
        pyatspi.Registry.deregisterEventListener(self._onMouseMoved, "mouse:abs")


    running = gobject.property(
        type=bool,
        nick='running',
        default=False,
        getter=_get_running,
        setter=_set_running)

    def _onMouseMoved(self, event):
        x, y = event.detail1, event.detail2
        freq, pan = self._calculateSoundParametersFromMousePosition(x, y)
        self._beep(freq, pan)

    def _calculateSoundParametersFromMousePosition(self, x, y):
        screen = gtk.gdk.screen_get_default()
        pan = 2 * (float(x) / screen.get_width())  -1
        freq = self.min_freq + (screen.get_height() - y) * (self.max_freq - self.min_freq) / screen.get_height()
        return (freq, pan)

    def _beep(self, frequency, panorama):
        self._player.set_state(gst.STATE_READY)
        self._source.set_property("freq", frequency)
        self._panorama.set_property("panorama", panorama)
        self._player.set_state(gst.STATE_PLAYING)

    def _onMessage(self, bus, message):
        t = message.type
        if t == gst.MESSAGE_EOS:
            self._player.set_state(gst.STATE_NULL)
        elif t == gst.MESSAGE_ERROR:
            self._player.set_state(gst.STATE_NULL)
            err, debug = message.parse_error()
            print "%s: %s" %(err, debug)


